add support for sd-event as event loop

defaultbranch opened the pull request:
Der grösste teil vom Code steckt in `sdevent.c`.

Als Vorlage habe ich [libuv.c](https://github.com/defaultbranch/libwebsockets/blob/main/lib/event-libs/libuv/libuv.c) und [libev.c](https://github.com/defaultbranch/libwebsockets/blob/main/lib/event-libs/libev/libev.c) verwendet.

Es gibt noch ein [README.md](https://github.com/defaultbranch/libwebsockets/blob/main/lib/event-libs/README.md) zum Einbinden von Event-Loop-Bibliotheken.

Commit:
mingw: libevent build fixes

lws-team committed:
h2: ignore unknown frame types a bit earlier

Chrome has started being able to issue frame type 0x42, we drop the connection
before we realize we wanted to ignore it.

This explicitly ignores it a bit earlier.

lws-team committed:
logging: reduce serving logs

Commit:
sspc: double free + rx ret not checked

Change-Id: I81b67cbb28591a20842fc7d6e09b78375845aae4

langchr86 commented on CMakeLists.txt:
```diff
@@ -152,6 +152,7 @@ option(LWS_WITH_LIBEV "Compile with support for libev" OFF)
 option(LWS_WITH_LIBUV "Compile with support for libuv" OFF)
 option(LWS_WITH_LIBEVENT "Compile with support for libevent" OFF)
 option(LWS_WITH_GLIB "Compile with support for glib event loop" OFF)
+option(LWS_WITH_SDEVENT "Compile with support for systemd event loop" OFF)
```
```suggestion
option(LWS_WITH_SDEVENT "Compile with support for systemd event loop: sd-event" OFF)
```

defaultbranch replied:
not sure, I wanted to keep the wording aligned with libuv/libevent/glib above.
Is "systemd event loop" ambiguous? I thought sd-event is the only systemd event loop?

langchr86 replied:
Yes it is the only one. I only think that we could easily miss the `d` in systemd and then think about raw Linux event loop or something like that. `sd-event` ist much more specific. Maybe we should use only `Compile with support for sd-event` instead of `systemd event loop`

defaultbranch replied:
Changed text to "Compile with support for sd-event"

langchr86 commented on include/libwebsockets/lws-context-vhost.h:
```diff
@@ -230,6 +230,9 @@
 	 * proxy client that is not trying to track system state by itself. */
 
+#define LWS_SERVER_OPTION_SDEVENT			 (1ll << 36)
+    /**< (CTX) Use systemd event loop */
```
```suggestion
    /**< (CTX) Use systemd event loop: sd-event */
```

defaultbranch replied:
(similar to the previous comment)

defaultbranch replied:
Changed to "Use sd-event loop"

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+#endif
+const lws_plugin_evlib_t evlib_sd = {
+        .hdr = {
+                "systemd event loop",
```
```suggestion
                "systemd event loop: sd-event",
```

defaultbranch replied:
(similar to the previous comment)

langchr86 reviewed (commented):
I miss one block of: https://github.com/defaultbranch/libwebsockets/blob/main/lib/event-libs/README.md

~~~~~~
Enabling event lib adoption

You need to add a LWS_SERVER_OPTION... flag as necessary in ./lib/libwebsockets.h enum lws_context_options, and follow the existing code in lws_create_context() to convert the flag into binding your ops struct to the context.
~~~~~~

langchr86 commented on CMakeLists-implied-options.txt:
```diff
@@ -87,6 +87,7 @@ if(LWS_WITH_DISTRO_RECOMMENDED)
 	set(LWS_WITH_LIBUV 1)				# libuv
 	set(LWS_WITH_LIBEV 1)				# libev
 	set(LWS_WITH_LIBEVENT 1)			# libevent
+	set(LWS_WITH_SDEVENT 0)				# sd-event
```
Why is this disabled by default?

defaultbranch replied:
I believe systemd is not as portable (it will probably never run on Windows), so I disable it by default.
My expectation would be that the libwebsockets maintainers will adjust this anyway.

langchr86 replied:
Make sense.

langchr86 commented on CMakeLists-implied-options.txt:
```diff
@@ -345,6 +347,10 @@ if (LWS_WITH_LIBEVENT)
 	set(LWS_WITH_LIBEVENT 1)
 endif()
 
+if (LWS_WITH_SDEVENT)
+	set(LWS_WITH_SDEVENT 1)
+endif()
```
The conditional and the set variable are both the exact same. What is this check intended for?

defaultbranch replied:
I copied this, not sure whether its a bug (could be), or it is there to ensure that anything "defined" is an integer 1 while anything undefined remains undefined (similar to what https://stackoverflow.com/questions/14751973/what-is-in-c does)

Need to check...

defaultbranch replied:
I forwarded the question to https://github.com/warmcat/libwebsockets/issues/2130, waiting for an answer.

defaultbranch replied:
Removed. (The maintainers found this pointless as well.)

langchr86 commented on lib/event-libs/sdevent/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+set(LIB_SYSTEMD_LIBRARIES CACHE PATH "Path to the libsystemd library")
+if ("${LWS_SYSTEMD_LIBRARIES}" STREQUAL "")
+    if (NOT LIB_SYSTEMD_FOUND)
+        message("searching for systemd")
```
I would remove this line because the `find_library` command will already print something similar.

defaultbranch replied:
Removed.

langchr86 commented on lib/event-libs/sdevent/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+
+if (LWS_WITH_EVLIB_PLUGINS)
+
+    create_evlib_plugin(evlib_sd sdevent.c ${LIBSYSTEMD_LIBRARIES})
```
Other implementations do also mention their private header here.

```suggestion
    create_evlib_plugin(evlib_sd sdevent.c private-lib-event-libs-sdevent.h ${LIBSYSTEMD_LIBRARIES})
```

defaultbranch replied:
Fixed.

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+#define pt_to_priv_sd(_pt) ((struct lws_pt_eventlibs_sdevent *)(_pt)->evlib_pt)
+#define wsi_to_priv_sd(_w) ((struct lws_wsi_watcher_sdevent *)(_w)->evlib_wsi)
+
+struct lws_context_eventlibs_sdevent {
+};
+
+struct lws_pt_eventlibs_sdevent {
+    struct lws_context_per_thread *pt;
+    struct sd_event *io_loop;
+    struct sd_event_source *sultimer;
+    struct sd_event_source *idletimer;
+};
+
+struct lws_vh_eventlibs_sdevent {
+};
+
+struct lws_wsi_watcher_sdevent {
+    struct sd_event_source *source;
+    uint32_t events;
+};
```
Those should go into `private-lib-event-libs-sdevent.h`

langchr86 commented on lib/event-libs/sdevent/private-lib-event-libs-sdevent.h:
```diff
@@ -0,0 +1,3 @@
+#include <private-lib-core.h>
```
Do we need to add the licence/copyright header in each of these source files?

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+        }
+        if (ptpriv->sultimer) {
+            sd_event_source_set_enabled(ptpriv->sultimer, SD_EVENT_OFF);
+            sd_event_source_unref(ptpriv->sultimer);
+            ptpriv->sultimer = NULL;
```
```suggestion
            ptpriv->sultimer = sd_event_source_unref(ptpriv->sultimer);
```

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+        }
+        if (ptpriv->idletimer) {
+            sd_event_source_set_enabled(ptpriv->sultimer, SD_EVENT_OFF);
+            sd_event_source_unref(ptpriv->sultimer);
+            ptpriv->sultimer = NULL;
```
```suggestion
            ptpriv->sultimer = sd_event_source_unref(ptpriv->sultimer);
```

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+            ptpriv->sultimer = NULL;
+        }
+        if (ptpriv->io_loop) {
+            sd_event_unref(ptpriv->io_loop);
+            ptpriv->io_loop = NULL;
```
```suggestion
            ptpriv->io_loop = sd_event_unref(ptpriv->io_loop);
```

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+            }
+            pt->event_loop_foreign = 0;
+        } else {
+            pt->event_loop_foreign = 1;
```
We need to increase the ref count here.

```suggestion
            pt->event_loop_foreign = 1;
            sd_event_ref(loop);
```

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+    // initialize accept/read for vhosts
+    // Note: default vhost usually not included here
+    for (struct lws_vhost *vh = context->vhost_list; vh; vh = vh->vhost_next) {
+        // call lws_event_loop_ops->init_vhost_listen_wsi
```
Is this a TODO? I suggest this line can be removed.

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+            return -1;
+        }
+
+        sd_event_source_set_enabled(ptpriv->idletimer, SD_EVENT_ON);
```
Check error code:

```suggestion
        if (0 > sd_event_source_set_enabled(ptpriv->idletimer, SD_EVENT_ON)) {
            return -1;
        }
```

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+            userdata
+    );
+
+    io_sd(wsi, LWS_EV_START | LWS_EV_READ);
```
As I understand the value `wsi_to_priv_sd(wsi)->events` is only set in this function. But it is already used in the line before `sd_event_add_io`. Is this correct?

langchr86 replied:
Or is this intended to afterwards set the events with `sd_event_source_set_io_events`? But then we could also pass `0` or a specific variable like `const int initial_empty_events = 0` in the `sd_event_add_io` call to make this more clear.

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+
+    void *userdata = wsi;
+
+    sd_event_add_io(
```
Missing error handling.

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+        if (!(wsi_to_priv_sd(wsi)->events & (EPOLLIN | EPOLLOUT)))
+            sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_ONESHOT);
+        else
+            sd_event_source_set_enabled(wsi_to_priv_sd(wsi)->source, SD_EVENT_OFF);
```
All `sd_event_` functions are missing error handling in this function.

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+
+        sd_event_source_set_io_events(wsi_to_priv_sd(wsi)->source, wsi_to_priv_sd(wsi)->events);
+
+        if (!(wsi_to_priv_sd(wsi)->events & (EPOLLIN | EPOLLOUT)))
```
Switch if/else clause to make this more understandable ( no `!` needed). Besides, is this even correct? If I understand correctly we do disable the event if at least one of the two event flags are set?

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+    eventfd.events = 0;
+    eventfd.revents = 0;
+
+    // TODO handle revents error bits
```
TODO

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+    }
+
+    lws_pt_unlock(pt);
+    lws_context_unlock(pt->context);
```
Is it correct to release the locks here? Because you still use `pt` later.

defaultbranch replied:
Copied from libuv/libev.

But I will check this again...

langchr86 commented on lib/event-libs/sdevent/sdevent.c:
```diff
@@ -0,0 +1,400 @@
+    lws_context_unlock(pt->context);
+
+    return -1;
+}
```
All sd_event_ functions are missing error handling in this function.

langchr86 reviewed (commented):
Looks not too bad. But I did not completely understand everything. Maybe I will with the next review after you added more tests. So that the code can mature. Most things I noticed is the missing error handling at many points.

lws-team committed:
async-dns: fix error path

lws-team committed:
libuv: context creation fails via destroy

When using a foreign libuv loop, context creation may fail after adding
handles to the foreign loop... if so, it can no longer deal with the
fatal error by unpicking the created context and returning NULL... it
has to brazen it out with a half-baked context that has already started
the destroy flow and allow the foreign loop to close out the handles
the usual way for libuv.

https://github.com/warmcat/libwebsockets/issues/2129

defaultbranch commented:
@langchr86 wrote:

> I miss one block of: https://github.com/defaultbranch/libwebsockets/blob/main/lib/event-libs/README.md
> 
> ```
> Enabling event lib adoption
> 
> You need to add a LWS_SERVER_OPTION... flag as necessary in ./lib/libwebsockets.h enum lws_context_options, and follow the existing code in lws_create_context() to convert the flag into binding your ops struct to the context.
> ```

This README is probably a bit out of date regarding the include file location. Today the enum and the new flag LWS_SERVER_OPTION_SDEVENT reside in include/libwebsockets/lws-context-vhost.h instead.

lws-team committed:
zip-fops: example: correct zipfile serving path

lws-team committed:
zip-fops: example: add concompressed mount too

lws-team committed:
h2: fix ignoring unknown frame payloads

lws-team committed:
zip-fops: uncompressed length fix

lws-team committed:
v4.1.6

lws-team committed:
esp32: update build arrangements

s-rogonov committed:
cmake: move LWS__INCLUDE_DIRS def to after REL_INCLIDE_DIR def

Generate the config file for the installation tree
fix wrong LWS__INCLUDE_DIRS definition

zishfish committed:
add support for sd-event as event loop

zishfish committed:
add sd-event support to example «minimal-http-server-eventlib-foreign»
